src/main/java/io/gingersnapproject/cdc/ManagedEngine.java

```diff
@@ -99,25 +116,83 @@ private static void stopEngine(StartStopEngine engine, String name) {
       }
    }
 
+   private static void shutdownEngine(StartStopEngine engine, String name) {
+      try {
+         engine.shutdown();
+      } catch (IOException e) {
+         log.error("Failed engine {} shutdown", name, e);
+         throw new RuntimeException(e);
+      }
+   }
+
    @Override
    public void addRule(String name, Rule rule) {
-      StartStopEngine sse = engines.computeIfAbsent(name, ignore ->
-            new StartStopEngine(new EngineWrapper(name, config, rule, cacheService, eventing))
-      );
-      sse.start();
+      knownRules.put(name, rule);
+      addRuleWithKnownMembers();
+   }
+
+   private void addRuleWithKnownMembers() {
+      AggregateCompletionStage<Void> stage = CompletionStages.aggregateCompletionStage();
+      for (Map.Entry<String, Rule> rule : knownRules.entrySet()) {
+         for (URI memberURI : knownMembers) {
+            var identifier = CacheIdentifier.of(rule.getKey(), memberURI);
+            engines.computeIfAbsent(identifier, ignore -> {
+               log.info("Creating new engine for {}", identifier);
+               var sse = new StartStopEngine(new EngineWrapper(identifier, config, rule.getValue(), cacheService, eventing));
+               stage.dependsOn(startEngine(identifier, sse));
+               return sse;
+            });
+         }
+      }
+      CompletableFutures.uncheckedAwait(stage.freeze().toCompletableFuture());
```

Comment:
Argh, now thinking of it, this could be dangerous in some scenarios. We retry forever in case of exceptions, and the poll executor is single-threaded, so:

1. We receive a join of a node (Unhealthy, for whatever reason);
2. The poll thread fires the join event, and we block here trying to connect to the unhealthy pod;
3. The pod leaves;
4. Thread is still blocked trying to connect.

I'll try to come up with something for this. I am afraid of concurrent join/leave to let this run without blocking.